func_server.cpp

```diff
@@ -29,59 +29,85 @@ Status FuncServiceImpl::unhook(ServerContext* context,
-    std::optional<std::string> replyMessage =
-        (warbleServer_.*
-         (warbleServer_.function_map_[event_function]))(request->payload());
-    if (replyMessage != std::nullopt &&
-        replyMessage.value() == "the user is not registered") {
-      // user is not registered
-      return Status(StatusCode::PERMISSION_DENIED, replyMessage.value());
-    }
-    if (replyMessage != std::nullopt &&
-        replyMessage.value() == "the reply id is invalid") {
-      // reply id is invalid
-      return Status(StatusCode::NOT_FOUND, replyMessage.value());
-    }
-    if (replyMessage != std::nullopt &&
-        replyMessage.value() == "read non-existing id") {
-      // reply id is invalid
-      return Status(StatusCode::NOT_FOUND, replyMessage.value());
-    }
-    if (replyMessage != std::nullopt && event_type == EVENT::REGISTER) {
-      return Status(StatusCode::ALREADY_EXISTS, replyMessage.value());
-    }
-    if (replyMessage != std::nullopt && event_type == EVENT::FOLLOW) {
-      if (replyMessage.value() == "cannot follow yourself" ||
-          replyMessage.value() ==
-              "cannot follow because your target is not registered") {
-        return Status(StatusCode::INVALID_ARGUMENT, replyMessage.value());
-      }
-      return Status(StatusCode::ALREADY_EXISTS, replyMessage.value());
+
+  if (event_function.empty()) {
+    return Status(StatusCode::NOT_FOUND, "the event is not hooked");
+  }
+
+  // potentially non terminating calls
+  if (event_type == EVENT::STREAM) {
+    int curr_loop = 0;
+    int loop_max = 5;
+    EventReply reply;
+    StreamReply streamReply;
+    while (curr_loop != loop_max) {
+      auto* payload = new google::protobuf::Any();
+
+      // make the call for latest streamed warble
+      std::optional<std::string> replyMessage =
+          (warbleServer_.*
+           (warbleServer_.function_map_[event_function]))(request->payload());
+
+      streamReply.ParseFromString(replyMessage.value());
+      payload->PackFrom(streamReply);
+      reply.set_allocated_payload(payload);
+      writer->Write(reply);
+      curr_loop++;
     }
-    if (event_type == EVENT::PROFILE) {
-      ProfileReply profileReply;
-      profileReply.ParseFromString(replyMessage.value());
-      payload->PackFrom(profileReply);
-    } else if (event_type == EVENT::READ) {
-      ReadReply readReply;
-      readReply.ParseFromString(replyMessage.value());
-      payload->PackFrom(readReply);
-    } else if (event_type == EVENT::WARBLE) {
-      WarbleReply warbleReply;
-      warbleReply.ParseFromString(replyMessage.value());
-      payload->PackFrom(warbleReply);
+    return Status::OK;
+  }
+
+  auto* payload = new google::protobuf::Any();
+  std::optional<std::string> replyMessage =
+      (warbleServer_.*
+       (warbleServer_.function_map_[event_function]))(request->payload());
+  if (replyMessage != std::nullopt &&
+      replyMessage.value() == "the user is not registered") {
+    // user is not registered
```

Comment:
check if the user is registered should be put in front of line 44

Reply:
check added